### PR TITLE
Flow tests with page objects

### LINKED ISSUE #10: Refactor flow tests to use page objects
#### As a developer I want my flow tests to be written using the object-oriented [page object pattern](http://martinfowler.com/bliki/PageObject.html).

__Story__:

Now that we have gotten some experience writing flow tests with Capybara,
we've noticed that these tests a) can be brittle in that changing selectors
requires updating many tests and b) use a low-level interface that can make
reading and writing flow tests more cumbersome than we'd like.

Fortunately, there is a better way: using page objects. With page objects,
we can create objects to encapsulate pages, elements, and actions that can be
taken in our application. We can design an interface for interacting with our
application that maps directly to how our users would think about interacting
with the application. This interface makes the tests easier to read, write,
and modify, and reduces the possibility of writing flaky or invalid tests.

Working on this story will be a bit different than your typical stories. We've
created a starting point for you in the form of a branch in your repo called
`flowTestsWithPageObjects`. This branch contains flow tests for adding images,
deleting images, and viewing images associated with a tag (in the
`images_crud_test.rb` file). These tests are fully written, but the page
objects they use are not. Your job is to fill in the page objects to make the
tests pass against your application.

The first step is to rebase the `flowTestsWithPageObjects` branch on top of
your updated master branch. Note that if you already have an
`images_crud_test.rb` file, there will be a merge conflict, so you may want to
rename your file first, rebase, make the tests on that branch pass, and then
migrate any other tests that were in your file to the page object style before
deleting your old test file.

Finally, there is a video to help you get started. You should watch this video
and also take a look at the ae_page_objects readme which will be an extremely
helpful resource as you learn to write page objects. Both are linked in the
reference material below.

__Acceptance criteria__:
- [ ] The flow tests have been refactored to use page objects.

__Discussion topics__:
- [ ] What makes a good page object interface?

__Reference material__:
- https://github.com/appfolio/ae_page_objects
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects

__Dependencies__:
- Delete Images
- Images by Tag

### PR DESCRIPTION
**Pull request by ptevis42**
Closes #10

### CODE REVIEW

**Comment by kkajla12** [comment]
Before I do a detailed review of this PR, I think we need to reorder and better group the changes. I would recommend that you group together your implementation of all of the page object classes first, then you can group together adding your page objects tests, and finally you can delete the existing flow tests. This will make the PR flow more naturally, which makes things easier for the reviewer as well as anyone who decides to look back on a PR in the future to understand what exactly you did.

**Comment by rmacklin** [comment]
@kkajla12 @ptevis42 The commits that came with the repository (the ones not by @ptevis42) should be left alone.

File: test/page_objects/images/image_card.rb [hunk 1]
```diff
@@ -1,12 +1,16 @@
 module PageObjects
   module Images
     class ImageCard < AePageObjects::Element
+      collection :tag_elements,
+                 locator: 'div.tag-list',
```
Comment (kkajla12) [inline test/page_objects/images/image_card.rb]: I don't think your selectors here need to be so specific. You have the `.tag-list` and `.tag` classes which can already help you uniquely select the tag list and each individual tag within the tag list using the `item_locator` attribute. I see that later on in b41a27fe68a3078519dca41ea949d2f1883de2d2 you add these classes (as `.js` classes), but maybe that change belongs in this commit.

File: test/page_objects/images/image_card.rb [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
+                 item_locator: 'span.tag'
```
Comment (kkajla12) [inline test/page_objects/images/image_card.rb]: see https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/4c90b73d0284ab6d9de7e10e6bcbd0c6ef34eadc#r117072981 above

File: test/page_objects/images/show_page.rb [hunk 3]
```diff
@@ -3,12 +3,16 @@ module Images
     class ShowPage < PageObjects::Document
       path :image
 
+      collection :tag_elements,
+                 locator: 'div.tag-list',
+                 item_locator: 'span.tag'
```
Comment (kkajla12) [inline test/page_objects/images/show_page.rb]: refer to https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/4c90b73d0284ab6d9de7e10e6bcbd0c6ef34eadc#r117072981 above

File: test/page_objects/images/show_page.rb [hunk 4]
```diff
@@ -3,12 +3,16 @@ module Images
...
+
       def image_url
-        # TODO
+        node.find('img.shared-image')[:src]
```
Comment (kkajla12) [inline test/page_objects/images/show_page.rb]: This selector might also be too specific

File: app/views/images/index.html.erb [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
+    <div class="col-lg-3 col-md-4 col-sm-6 col-xs-12 js-image-card-container">
+      <div class="card">
+        <a href="<%= image_path(image) %>" class="js-image-show-link">
           <%= image_tag image.url, class: 'img-fluid' %>
         </a>
-        <div class="tag-list">
+        <div class="js-tag-list">
           <% image.tags.each do |tag| %>
-            <span class="tag tag-default"><%= link_to tag.name, images_path(tag: tag.name) %></span>
+            <span class="tag tag-default js-tag"><%= link_to tag.name, images_path(tag: tag.name) %></span>
```
Comment (kkajla12) [inline app/views/images/index.html.erb]: As I mentioned elsewhere, the addition of the js classes on the tag elements really belong in your add test commit because they are related to that commit.

File: app/views/images/show.html.erb [hunk 6]
```diff
@@ -1,9 +1,9 @@
 <div class="image-container">
   <%= image_tag @image.url, class: 'shared-image'%>
 </div>
-<div class="tag-list">
+<div class="js-tag-list">
   <% @image.tags.each do |tag| %>
-    <span class="tag tag-default"><%= link_to tag.name, images_path(tag: tag.name) %></span>
+    <span class="tag tag-default js-tag"><%= link_to tag.name, images_path(tag: tag.name) %></span>
```
Comment (kkajla12) [inline app/views/images/show.html.erb]: refer to https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/b41a27fe68a3078519dca41ea949d2f1883de2d2#r117075254

File: app/views/images/show.html.erb [hunk 7]
```diff
@@ -1,9 +1,9 @@
 <div class="image-container">
   <%= image_tag @image.url, class: 'shared-image'%>
 </div>
-<div class="tag-list">
+<div class="js-tag-list">
```
Comment (kkajla12) [inline app/views/images/show.html.erb]: refer to https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/b41a27fe68a3078519dca41ea949d2f1883de2d2#r117075254

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -54,11 +54,11 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     image.save!
     get images_path
     assert_response :ok
-    assert_select 'div.tag-list' do
-      assert_select 'span.tag' do
+    assert_select 'div.js-tag-list' do
```
Comment (kkajla12) [inline test/controllers/images_controller_test.rb]: I think the changes to this file that result from the css class changes either belong in their own commit, or with your add test commit.

File: test/page_objects/images/index_page.rb [hunk 9]
```diff
@@ -4,11 +4,12 @@ class IndexPage < PageObjects::Document
       path :images
 
       collection :images,
-                 locator: '#images_list',
+                 locator: '.js-images-list',
```
Comment (kkajla12) [inline test/page_objects/images/index_page.rb]: I think it's good that you're being consistent here

File: test/page_objects/images/index_page.rb [hunk 10]
```diff
@@ -4,11 +4,12 @@ class IndexPage < PageObjects::Document
...
                  item_locator: '.js-image-card-container',
                  contains: ImageCard do
         def view!
-          # TODO
+          node.find('a.js-image-show-link').click
```
Comment (kkajla12) [inline test/page_objects/images/index_page.rb]: css selector is too specific.

File: test/page_objects/images/show_page.rb [hunk 11]
```diff
@@ -4,8 +4,8 @@ class ShowPage < PageObjects::Document
       path :image
 
       collection :tag_elements,
-                 locator: 'div.tag-list',
-                 item_locator: 'span.tag'
+                 locator: '.js-tag-list',
```
Comment (kkajla12) [inline test/page_objects/images/show_page.rb]: for these changes, refer to https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/b41a27fe68a3078519dca41ea949d2f1883de2d2#r117075254

File: test/flows/images_crud_test.rb [hunk 12]
```diff
@@ -28,24 +28,22 @@ class ImagesCrudTest < FlowTestCase
   test 'delete an image' do
     cute_puppy_url = 'http://ghk.h-cdn.co/assets/16/09/980x490/landscape-1457107485-gettyimages-512366437.jpg'
     ugly_cat_url = 'http://www.ugly-cat.com/ugly-cats/uglycat041.jpg'
-    Image.create!([
-       { url: cute_puppy_url, tag_list: 'puppy, cute' },
-       { url: ugly_cat_url, tag_list: 'cat, ugly' }
-    ])
+    Image.create!([{ url: cute_puppy_url, tag_list: 'puppy, cute' },
```
Comment (kkajla12) [inline test/flows/images_crud_test.rb]: The changes in this file that continue implementing the delete test belong in the previous commit with all of the related changes

File: app/views/images/show.html.erb [hunk 13]
```diff
@@ -1,11 +1,11 @@
-<div class="image-container">
-  <%= image_tag @image.url, class: 'shared-image'%>
+<div class="js-image-container">
```
Comment (kkajla12) [inline app/views/images/show.html.erb]: Did the `image-container` class have any styling associated with it? If so, you shouldn't be changing it to `js-image-container` because you'll lose all of that styling then. If not, then the `image-container` class should instead be an id and then you can continue to use it or this change is ok.

File: app/views/images/show.html.erb [hunk 14]
```diff
@@ -1,11 +1,11 @@
...
+  <%= image_tag @image.url, class: 'js-shared-image'%>
```
Comment (kkajla12) [inline app/views/images/show.html.erb]: refer to https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/869456729737b4cf62942a8cc57c2c2c5429e336#r117077723

File: app/views/images/show.html.erb [hunk 15]
```diff
@@ -1,11 +1,11 @@
...
 </div>
 <div class="js-tag-list">
   <% @image.tags.each do |tag| %>
     <span class="tag tag-default js-tag"><%= link_to tag.name, images_path(tag: tag.name) %></span>
   <% end %>
 </div>
-<div class="action-list">
+<div class="js-action-list">
```
Comment (kkajla12) [inline app/views/images/show.html.erb]: refer to https://github.com/AppFolioOnboarding/image-sharer-ptevis42/pull/27/commits/869456729737b4cf62942a8cc57c2c2c5429e336#r117077723
